src/setting_sbp_cb.c

```diff
@@ -0,0 +1,401 @@
+/*
+ * Copyright (C) 2019 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <libsbp/settings.h>
+
+#include <libsettings/settings_util.h>
+
+#include <internal/request_state.h>
+#include <internal/setting_data.h>
+#include <internal/setting_def.h>
+#include <internal/setting_sbp_cb.h>
+
+/**
+ * @brief setting_send_write_response
+ * @param write_response: pre-formatted write response sbp message
+ * @param len: length of the message
+ * @return zero on success, -1 if message failed to send
+ */
+static int setting_send_write_response(settings_t *ctx,
+                                       msg_settings_write_resp_t *write_response,
+                                       uint8_t len)
+{
+  settings_api_t *api = &ctx->client_iface;
+  if (api->send(api->ctx, SBP_MSG_SETTINGS_WRITE_RESP, len, (uint8_t *)write_response) != 0) {
+    api->log(log_err, "sending settings write response failed");
+    return -1;
+  }
+  return 0;
+}
+
+static void setting_update_value(settings_t *ctx, uint8_t *msg, uint8_t len)
+{
+  /* Extract parameters from message:
+   * 4 null terminated strings: section, name, value and type.
+   * Expect to find at least section, name and value.
+   */
+  const char *section = NULL, *name = NULL, *value = NULL, *type = NULL;
+  if (settings_parse((char *)msg, len, &section, &name, &value, &type) < SETTINGS_TOKENS_VALUE) {
+    ctx->client_iface.log(log_warning, "settings register resp cb, error parsing setting");
+    return;
+  }
+
+  /* Look up setting data */
+  setting_data_t *setting_data = setting_data_lookup(ctx->setting_data_list, section, name);
+  if (setting_data == NULL) {
+    return;
+  }
+
+  if (setting_data->watchonly) {
+    return;
+  }
+
+  settings_write_res_t write_result = setting_data_update_value(setting_data, value);
+
+  uint8_t resp[SETTINGS_BUFLEN] = {0};
+  uint8_t resp_len = 0;
+  msg_settings_write_resp_t *write_response = (msg_settings_write_resp_t *)resp;
+  write_response->status = write_result;
+  resp_len += sizeof(write_response->status);
+  int l = setting_data_format(setting_data,
+                              false,
+                              write_response->setting,
+                              sizeof(resp) - resp_len,
+                              NULL);
+  if (l < 0) {
+    return;
+  }
+  resp_len += l;
+
+  setting_send_write_response(ctx, write_response, resp_len);
```

Comment:
If I followed the code correctly, this is needed for updating watched values (in case of registration). Maybe `SBP_MSG_SETTINGS_UPDATE_EVENT` would be more suitable? It would leave room for future purposes for example if we ever have a case where the device is forced to change a setting value due to whatever internal reasons.

Reply:
agreed, I think we need to redesign the watcher behavior before we can update this behavior to it's proper form. for now this is functionally equivalent to the current design so let's go ahead and merge then iterate.